exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java

```diff
@@ -1221,6 +1223,12 @@ public void interrupted(final InterruptedException e) {
      *   to the user
      */
     public void moveToState(final QueryState newState, final Exception ex) {
+      // if the current thread is the foreman thread, throw an exception
+      // otherwise the foreman will be blocked forever on acceptExternalEvents
+      if (myThreadRef == Thread.currentThread()) {
```

Comment:
This won't work when submitting remote fragments as the moveToState method can be called either by the foreman thread or by an rpc thread.

Anyway, I'm working on a better fix that would remove the latch completely.